tasks/libs/ciproviders/gitlab_api.py

```diff
@@ -84,25 +114,150 @@ def __init__(self, references):
 
     @classmethod
     def from_yaml(cls, loader, node):
-        return UserList(loader.construct_sequence(node))
+        return YamlReferenceTagList(loader.construct_sequence(node))
 
     @classmethod
     def to_yaml(cls, dumper, data):
-        return dumper.represent_sequence(cls.yaml_tag, data.data, flow_style=True)
+        return dumper.represent_sequence(cls.yaml_tag, data, flow_style=True)
+
+
+def convert_to_config_node(node):
+    """
+    Convert json data to ConfigNode
+    """
+    if isinstance(node, dict):
+        return ConfigNodeDict({k: convert_to_config_node(v) for k, v in node.items()})
+    elif isinstance(node, list):
+        constructor = YamlReferenceTagList if isinstance(node, YamlReferenceTagList) else ConfigNodeList
+
+        return constructor([convert_to_config_node(v) for v in node])
+    else:
+        return node
+
+
+def apply_yaml_extends(config: dict, node):
+    """
+    Applies `extends` yaml tags to the node and its children inplace
+
+    > Example:
+    Config:
+    ```yaml
+    .parent:
+        hello: world
+    node:
+        extends: .parent
+    ```
+
+    apply_yaml_extends(node) updates node to:
+    ```yaml
+    node:
+        hello: world
+    ```
+    """
+    # Ensure node is an object that can contain extends
+    if not isinstance(node, dict):
+        return
+
+    if 'extends' in node:
+        parents = node['extends']
+        if isinstance(parents, str):
+            parents = [parents]
 
+        # Merge parent
+        for parent_name in parents:
+            parent = config[parent_name]
+            apply_yaml_postprocessing(config, parent)
+            for key, value in parent.items():
+                if key not in node:
+                    node[key] = value
 
-def generate_gitlab_full_configuration(input_file, context=None, compare_to=None):
+        del node['extends']
+
+
+def apply_yaml_reference(config: dict, node):
+    """
+    Applies `!reference` gitlab yaml tags to the node and its children inplace
+
+    > Example:
+    Config:
+    ```yaml
+    .colors:
+        - red
+        - green
+        - blue
+    node:
+        colors: !reference [.colors]
+    ```
+
+    apply_yaml_extends(node) updates node to:
+    ```yaml
+    node:
+        colors:
+            - red
+            - green
+            - blue
+    ```
+    """
+
+    def apply_ref(value):
+        """
+        Applies reference tags
+        """
+        if isinstance(value, YamlReferenceTagList):
+            assert value != [], 'Empty reference tag'
+
+            # [a, b, c] -> config[a][b][c]
+            ref_value = config[value[0]]
```

Comment:
what if the referenced value is not a `list` (I think it might be possible)

Reply:
It is not possible, otherwise the yaml is invalid. I tried with the validate endpoint. Furthermore it says in the [documentation](https://docs.gitlab.com/ee/ci/yaml/yaml_optimization.html#configure-your-ide-to-support-reference-tags) that it should be a sequence